Map the world entity managers & surrounding classes

hedgehog1029 opened the pull request:
Continuation & update of #1851.

- `ServerEntityManager`: server's entity management, tracks entities in each chunk
- `ClientEntityManager`: equivalent for client
- `EntityLike`: what it says on the tin
- `EntityTracker`: given to each entity as a way to call back in to the entity manager
- `EntityLoadingStage`: stage of loading entity from disk
- `TypeFilter`: downcasting utility
- `ChunkDataAccess`: interface for specific access to chunk data, future-based
- `ChunkDataList`: very generic list wrapper which includes a position (ChunkPos)
- `ChunkStatusChangeListener`: callback for when the chunk status changes, i guess
- `EntityChunkStatus`: encodes whether an entity should be ticked and/or be tracked by the manager
- `EntityIdMapSet`, `EntitySet`, etc.: it stores da entities by id efficiently
- `EntityTrackingGroup`: group of tracked entities treated as a blob in terms of `EntityChunkStatus` and such
- `SectionedEntityCache`: stores tracking groups keyed by subchunk position and does a Whole bunch of stuff with that

lotsa very generic classes here tbh

kvverti committed:
Map a lot of entity world interaction shenanigans. Feeling tired might remap later.

kvverti committed:
Merge branch '20w45a' of github.com:FabricMC/yarn into world-entity

kvverti committed:
Map more entity loading stuff.

kvverti committed:
Map ChunkStatusChangeListener.

kvverti committed:
Filter -> TypeFilter

kvverti committed:
Some parameters and mappings.

kvverti committed:
Some docs and mapping of methods called by the entity stuff.

kvverti committed:
Update mappings from review.

kvverti committed:
Merge branch '20w45a' into world-entity

kvverti committed:
listener -> handler

hedgehog1029 committed:
Merge 20w49a into world-entity

hedgehog1029 committed:
Apply some suggestions & map extra names

Gegy commented on mappings/net/minecraft/world/entity/EntityTrackingGroup.mapping:
```diff
@@ -0,0 +1,24 @@
+CLASS net/minecraft/class_5572 net/minecraft/world/entity/EntityTrackingGroup
```
Group is a very loose description for this class, given its usage to specifically represent a section containing entities. It also holds state relating to the chunk that contains it.

hedgehog1029 replied:
I like to think of it as a "group" of entities that are processed together by the manager; is there an established naming scheme for these already?

Gegy replied:
By their usage, they are used to represent chunk sections for entities. Especially given they reference chunk state it makes sense to name according to that.

hedgehog1029 replied:
> It also holds state relating to the chunk that contains it.

it occurs to me that the problem name here might actually be `EntityChunkStatus`, which is actually a flags enum encoding whether the entities in this group should be ticked, tracked, or neither. the "state" that you mention isn't related to the parent chunk, it's exclusively related to the group

Gegy replied:
The issue for me comes by reading the name "Group", I can have no idea along what lines entities are going to be grouped. It seems totally arbitrary- and hence any code working with these becomes unclear. You will always be working with a "section" of entities through this class.

hedgehog1029 replied:
OK - sounds good to me. `EntityTrackingSection` then?

Gegy commented on mappings/net/minecraft/entity/Entity.mapping:
```diff
@@ -257,15 +259,7 @@ CLASS net/minecraft/class_1297 net/minecraft/entity/Entity
 	METHOD method_31481 isRemoved ()Z
 	METHOD method_31482 unsetRemoved ()V
 	METHOD method_31483 getFirstPassenger ()Lnet/minecraft/class_1297;
-	METHOD method_32312 getFrozenTicks ()I
```
Merge issues? Should these have been removed?

hedgehog1029 replied:
huh weird

hedgehog1029 replied:
Restored in a6fe16a9c

Gegy commented on mappings/net/minecraft/entity/EntityLike.mapping:
```diff
@@ -0,0 +1,20 @@
+	METHOD method_31744 setEntityTracker (Lnet/minecraft/class_5569;)V
+		COMMENT Sets the behavior to execute when an entity's state updates.
+		COMMENT
+		COMMENT @see EntityChangeHandler
```
Comment reference needs to be renamed

hedgehog1029 replied:
Resolved in a6fe16a9c

Gegy commented on mappings/net/minecraft/server/world/ServerEntityManager.mapping:
```diff
@@ -0,0 +1,98 @@
+		ARG 1 writer
+	METHOD method_31827 containsEntity (Ljava/util/UUID;)Z
+		ARG 1 entityUuid
+	METHOD method_31828 addEntitiesWithFlagSet (Ljava/util/stream/Stream;)V
```
This naming `WithFlagSet` and `WithFlagCleared` is very vague, and probably not worth including unless using a proper name. As far as I can tell, this flag looks to relate to whether the entities are being loaded from a "legacy" world chunk that contains entities, or if it is being loaded from the new entity chunks.

hedgehog1029 replied:
inherited from the previous PR and I agree, the flag doesn't actually do anything at the moment so it's difficult to name the methods surrounding it

hedgehog1029 replied:
Resolved in a6fe16a9c

Gegy commented on mappings/net/minecraft/world/chunk/ChunkDataList.mapping:
```diff
@@ -0,0 +1,9 @@
+CLASS net/minecraft/class_5566 net/minecraft/world/chunk/ChunkDataList
```
Data is already plural, would this make sense as just `ChunkData`?

hedgehog1029 replied:
this class is so generic i hate it. `ChunkPositionData` maybe? it's literally a wrapper around a list with a `ChunkPos` attached

Gegy replied:
It is data associated with a chunk, I think the name is probably fine? ChunkPositionData might imply that the data *is* the position, which may be a bit confusing?

kvverti replied:
`ChunkData` implies a single blob of data. This is a list of blobs of data.

Gegy commented on mappings/net/minecraft/world/chunk/ChunkDataAccess.mapping:
```diff
@@ -0,0 +1,6 @@
+CLASS net/minecraft/class_5571 net/minecraft/world/chunk/ChunkDataAccess
```
Access doesn't have any specific implications about where the data is coming from, and it might be worth clarifying that it loads from disk in the name?

hedgehog1029 replied:
do we have a naming scheme for "loads from disk"? `Io` maybe?

Gegy replied:
I am pretty sure that has been used in other places, yeah

kvverti replied:
`ChunkDataIO` is good if it's not taken (do we have `ChunkIO` already?)

Gegy commented on mappings/net/minecraft/world/entity/EntityIdMap.mapping:
```diff
@@ -0,0 +1,17 @@
+CLASS net/minecraft/class_5577 net/minecraft/world/entity/EntityIdMap
```
This works with more than just IDs, I think this name may be too specific?

hedgehog1029 replied:
it deals in IDs and UUIDs, which I argue are a type of IDs. no idea what to name this class and its implementor though, it's all too generic

Gegy replied:
The alternative names proposed in the former PR were `EntityStore`/`SectionEntityStore`, which are not ideal, but I think at least a bit less confusing. I suppose the issue with the name `EntityIdMap` is that it implies that it works solely with IDs- it's not clear what it is mapping those IDs to.

Gegy commented on mappings/net/minecraft/world/entity/EntityIdMapSet.mapping:
```diff
@@ -0,0 +1,7 @@
+CLASS net/minecraft/class_5578 net/minecraft/world/entity/EntityIdMapSet
```
`MapSet` is very confusing. This is an implementation of the parent interface that stores entities indexed by section, the name should probably clarify that.

hedgehog1029 replied:
i hate this class, it's basically just a read-only view of two other data classes

`EntityMapView`? its methods are *mostly* map-like (fetch by id, iterate over values)

Gegy replied:
I'd propose `Sectioned` prefix to the parent interface given its implementation.

Gegy commented on mappings/net/minecraft/world/entity/SectionedEntityCache.mapping:
```diff
@@ -0,0 +1,45 @@
+CLASS net/minecraft/class_5573 net/minecraft/world/entity/SectionedEntityCache
```
This doesn't behave as just a cache, I think that description might be misleading.

hedgehog1029 replied:
I suspect kvverti named it `Cache` because of all the fastutil data structures backing it up, kind of implies speed. this is the class that tracks entities in groups of subchunks at runtime, so idk

Gegy commented on mappings/net/minecraft/world/entity/SectionedEntityCache.mapping:
```diff
@@ -0,0 +1,45 @@
+		ARG 3 action
+	METHOD method_31775 intersecting (Lnet/minecraft/class_238;)Ljava/util/function/Predicate;
+		ARG 0 box
+	METHOD method_31777 forEachInBox (Lnet/minecraft/class_238;Ljava/util/function/Consumer;)V
```
What's the difference between `InBox` and `Intersects`? Also, this works with sections specifically whereas some of the other functions don't, they should be clarified as different.

hedgehog1029 replied:
`Insersects` uses `Box#intersects` and is filtered from `InBox` which does its own checking logic

Gegy replied:
But do they not perform the same action? Also there is still the issue that the functions do not clarify whether they are working with entities or groups/sections

kvverti replied:
One executes a callback, one returns a predicate .. they are nothing alike.

kvverti replied:
`forEachIntersecting` would be acceptable

i509VCB commented on mappings/net/minecraft/util/thread/MessageListener.mapping:
```diff
@@ -7,3 +7,5 @@ CLASS net/minecraft/class_3906 net/minecraft/util/thread/MessageListener
 		ARG 1 action
 	METHOD method_17345 ask (Ljava/util/function/Function;)Ljava/util/concurrent/CompletableFuture;
 		ARG 1 messageProvider
+	METHOD method_27918 askFallible (Ljava/util/function/Function;)Ljava/util/concurrent/CompletableFuture;
```
askFallible as in?

hedgehog1029 replied:
seems this is near-identical to `ask` but it can return a `CompletableFuture` that completes exceptionally, wheras `ask` will never(?) do that

i509VCB replied:
It would be nice if you could make that distinguishment in the javadoc

hedgehog1029 replied:
Added in e9f0cea

hedgehog1029 committed:
Apply some suggestions to clean up

- Restored missing Entity methods
- Fixed comment reference
- Unmapped generic `flagSet`-named methods

hedgehog1029 committed:
Add javadocs for MessageListener

liach commented on mappings/net/minecraft/entity/EntityLike.mapping:
```diff
@@ -0,0 +1,20 @@
+CLASS net/minecraft/class_5568 net/minecraft/entity/EntityLike
+	COMMENT A superinterface of {@link Entity}.
```
This thing feels like those world interfaces. To me this is like "savable entity" that dedicates to s/l operations

YanisBft commented on mappings/net/minecraft/server/world/ServerEntityManager.mapping:
```diff
@@ -0,0 +1,98 @@
+		COMMENT Attempts to write the entity data for entities at the given {@link ChunkPos}, and runs the given
+		COMMENT action for each entity after saving.
+		COMMENT
+		COMMENT @return True if the entities were written, false if they were not.
```
```suggestion
		COMMENT @return {@code true} if the entities were written, {@code false} if they were not
```

YanisBft commented on mappings/net/minecraft/server/world/ServerEntityManager.mapping:
```diff
@@ -0,0 +1,98 @@
+		ARG 1 entities
+	METHOD method_31836 flushChunks ()V
+	METHOD method_31837 tryUnloadChunk (J)Z
+		COMMENT Attempts to save the chunk, unloading all entities if successful
```
```suggestion
		COMMENT Attempts to save the chunk, unloading all entities if successful.
```

YanisBft commented on mappings/net/minecraft/util/TypeFilter.mapping:
```diff
@@ -0,0 +1,10 @@
+		COMMENT Creates a filter whose filtering condition is whether the object is an instance of the given class.
+		ARG 0 cls
+	METHOD method_31796 downcast (Ljava/lang/Object;)Ljava/lang/Object;
+		COMMENT If the argument can be converted to the type {@code T}, return the argument, otherwise {@code null}.
```
```suggestion
		COMMENT Checks if the argument can be converted to the type {@code T} and returns the argument, or {@code null} otherwise.
```

haykam821 commented on mappings/net/minecraft/entity/EntityLike.mapping:
```diff
@@ -0,0 +1,20 @@
+CLASS net/minecraft/class_5568 net/minecraft/entity/EntityLike
```
```suggestion
CLASS net/minecraft/class_5568 net/minecraft/entity/EntityTrackable
```

YanisBft commented on mappings/net/minecraft/util/thread/MessageListener.mapping:
```diff
@@ -6,4 +6,18 @@ CLASS net/minecraft/class_3906 net/minecraft/util/thread/MessageListener
 		ARG 0 name
 		ARG 1 action
 	METHOD method_17345 ask (Ljava/util/function/Function;)Ljava/util/concurrent/CompletableFuture;
+		COMMENT Ask a message provider for a message.
```
```suggestion
		COMMENT Asks a message provider for a message.
```

YanisBft commented on mappings/net/minecraft/util/thread/MessageListener.mapping:
```diff
@@ -6,4 +6,18 @@ CLASS net/minecraft/class_3906 net/minecraft/util/thread/MessageListener
+		COMMENT @return CompletableFuture future that completes with the received message
+		ARG 1 messageProvider
+	METHOD method_27918 askFallible (Ljava/util/function/Function;)Ljava/util/concurrent/CompletableFuture;
+		COMMENT Ask a falliable message provider for a message.
```
```suggestion
		COMMENT Asks a fallible message provider for a message.
```

YanisBft commented on mappings/net/minecraft/world/entity/EntitySet.mapping:
```diff
@@ -0,0 +1,18 @@
+	FIELD field_27244 LOGGER Lorg/apache/logging/log4j/Logger;
+	FIELD field_27245 entitiesById Lit/unimi/dsi/fastutil/ints/Int2ObjectMap;
+	FIELD field_27246 entitiesByUuid Ljava/util/Map;
+	METHOD method_31751 entities ()Ljava/lang/Iterable;
```
```suggestion
	METHOD method_31751 iterate ()Ljava/lang/Iterable;
```

YanisBft commented on mappings/net/minecraft/entity/Entity.mapping:
```diff
@@ -36,6 +37,7 @@ CLASS net/minecraft/class_1297 net/minecraft/entity/Entity
 	FIELD field_5979 passengerList Lcom/google/common/collect/ImmutableList;
 	FIELD field_5981 uuidString Ljava/lang/String;
 	FIELD field_5982 prevYaw F
+	FIELD field_5983 teleporting Z
```
```suggestion
```
Doesn't exist

YanisBft commented on mappings/net/minecraft/entity/Entity.mapping:
```diff
@@ -508,6 +511,9 @@ CLASS net/minecraft/class_1297 net/minecraft/entity/Entity
 	METHOD method_5756 isInvisibleTo (Lnet/minecraft/class_1657;)Z
 		ARG 1 player
 	METHOD method_5757 isInsideWall ()Z
+	METHOD method_5758 equip (ILnet/minecraft/class_1799;)Z
+		ARG 1 slot
+		ARG 2 item
```
```suggestion
```

YanisBft commented on mappings/net/minecraft/entity/Entity.mapping:
```diff
@@ -678,6 +684,8 @@ CLASS net/minecraft/class_1297 net/minecraft/entity/Entity
 	METHOD method_5864 getType ()Lnet/minecraft/class_1299;
 	METHOD method_5865 updatePassengerPosition (Lnet/minecraft/class_1297;)V
 		ARG 1 passenger
+	METHOD method_5866 setWorld (Lnet/minecraft/class_1937;)V
+		ARG 1 world
```
```suggestion
```

YanisBft commented on mappings/net/minecraft/entity/EntityLike.mapping:
```diff
@@ -0,0 +1,20 @@
+CLASS net/minecraft/class_5568 net/minecraft/entity/EntityLike
```
```suggestion
CLASS net/minecraft/class_5568 net/minecraft/entity/EntityAccess
```
Can be another solution

liach replied:
Per our world naming guide we'd name it like `StoredEntity` `PersistentEntity` etc

YanisBft commented on mappings/net/minecraft/world/entity/EntityIdMap.mapping:
```diff
@@ -0,0 +1,17 @@
+CLASS net/minecraft/class_5577 net/minecraft/world/entity/EntityIdMap
+	COMMENT An interface exposing access to stored entities.
+	METHOD method_31803 entities ()Ljava/lang/Iterable;
```
```suggestion
	METHOD method_31803 iterate ()Ljava/lang/Iterable;
```

YanisBft commented on mappings/net/minecraft/world/entity/EntityIdMapSet.mapping:
```diff
@@ -0,0 +1,7 @@
+CLASS net/minecraft/class_5578 net/minecraft/world/entity/EntityIdMapSet
+	COMMENT
```
```suggestion
```

YanisBft commented on mappings/net/minecraft/world/entity/EntityIdMapSet.mapping:
```diff
@@ -0,0 +1,7 @@
+	FIELD field_27258 entities Lnet/minecraft/class_5570;
+	FIELD field_27259 sectionedEntities Lnet/minecraft/class_5573;
+	METHOD <init> (Lnet/minecraft/class_5570;Lnet/minecraft/class_5573;)V
+		ARG 1 entityStore
```
```suggestion
		ARG 1 entities
```

YanisBft commented on mappings/net/minecraft/world/entity/EntityTrackingGroup.mapping:
```diff
@@ -0,0 +1,24 @@
+	FIELD field_27248 backingList Lnet/minecraft/class_3509;
+	FIELD field_27249 chunkStatus Lnet/minecraft/class_5584;
+	METHOD <init> (Ljava/lang/Class;Lnet/minecraft/class_5584;)V
+		ARG 1 cls
```
```suggestion
		ARG 1 entityClass
```

hedgehog1029 committed:
Apply some cleanup from code review

Also removes some bits not present in 20w51a

Co-authored-by: YanisBft <doublecraft.official@gmail.com>

YanisBft commented:
Would be nice to get this done before the next snapshot

YanisBft commented:
Has been superseded by #1967
